design/decisions/0008-extensible-config-management.md

```diff
@@ -0,0 +1,210 @@
+  in the format <config_type>::<plugin_type>.
+  For example, if a `mosquitto` plugin lists two different config types: `mosquitto.conf` and `mosquitto.acl`,
+  both types would be reported as `mosquitto.conf::mosquitto` and `mosquitto.acl::mosquitto`.
+- When a `config_snapshot` request is received for a type, call the `get` command of the corresponding plugin for that type
+  which can be derived from the `::` suffix of that type.
+  The `type` passed to the `get` command would not include the plugin suffix.
+  If there is no explicit suffix, it is delegated to the default file plugin.
+- When a `config_update` request is received for a type, the following actions are performed in sequence:
+  - Call `get` command of the corresponding plugin and cache the target temporary file as a backup.
+  - Call `set` command of the corresponding plugin with the updated config path in the argument.
+  - Call `finalize` command of the plugin and complete the operation if successful.
+  - If `set` or `finalize` commands fail, call the `set` command again with the original config file backup as the target.
+  - Call `finalize` again if the previous `set` attempt to restore teh original configuration succeeded and fail the operation.
+
+### Phase 2
+
+Instead of the `exeucting` phase of `config_update` workflow doing everything in that single step,
+break it down into multiple smaller stages that corresponds to each plugin command/phase as well as follows:
+
+```config_update.toml
+operation = "config_update"
+
+[init]
+action = "proceed"
+on_success = "scheduled"
+
+[scheduled]
+action = "proceed"
+on_success = "executing"
+
+[executing]
+action = "builtin:config_update:executing"
+on_success = "download"
+
+[download]
+action = "builtin:download"
+on_success = "prepare"
+
+[prepare]
+action = "builtin:backup_file"
+on_success = "apply"
+
+[apply]
+action = "builtin:set_file"
+on_success = "validate"
+
+[validate]
+action = "builtin:config_update:validate"
+on_success = "commit"
+on_error = "rollback"
+
+[commit]
+action = "builtin:config_update:apply"
+on_success = "successful"
+
+[rollback]
+action = "builtin:config_update:rollback"
+on_success = "failed"
+
+[successful]
+action = "cleanup"
+
+[failed]
+action = "cleanup"
+```
+
+This helps customers override a single stage like `download` to do it their way,
+while still reusing the rest of the functionality as-is.
+
+### Unresolved problems
+
+1. When a new piece of software is installed on the device,
+   which mostly would have come with its own configurations and log files,
+   how to "notify tedge" so that it refreshes its
+   - updated software list
+   - supported config list
+   - updated logs list
+2. Even if a REST API or inotify API to refresh the agent is provided,
+   who would trigger this when a software is installed by various installation methods like:
+   - software packages by distros
```

Comment:
inotify would be triggered for any software packages as they generally change files on the system, so less of an issue with this scenario.